lib/nodetracker/implementation/tracker_test.go

```diff
@@ -0,0 +1,234 @@
+		addr        string
+	}{
+		{
+			nodeID:      "node-1",
+			proxyID:     "proxy-1",
+			clusterName: "cluster",
+			addr:        "proxy-1:3080",
+		},
+		{
+			nodeID:      "node-1",
+			proxyID:     "proxy-2",
+			clusterName: "cluster",
+			addr:        "proxy-2:3080",
+		},
+		{
+			nodeID:      "node-2",
+			proxyID:     "proxy-2",
+			clusterName: "cluster",
+			addr:        "proxy-2:3080",
+		},
+		{
+			nodeID:      "node-2",
+			proxyID:     "proxy-3",
+			clusterName: "cluster",
+			addr:        "proxy-3:3080",
+		},
+	}
+
+	totalStates := 3
+	states := []struct {
+		nodeID               string
+		proxiesAfterEachTick []int
+	}{
+		{
+			nodeID:               "node-1",
+			proxiesAfterEachTick: []int{2, 0, 0},
+		},
+		{
+			nodeID:               "node-2",
+			proxiesAfterEachTick: []int{2, 1, 1},
+		},
+	}
+
+	var tracker api.Tracker
+	ctx := context.Background()
+	clock := clockwork.NewFakeClock()
+
+	// the callback is used to check the results after each tick of the
+	// proxy relationship cleaner
+	var wg sync.WaitGroup
+	wg.Add(totalStates)
+	stateCounter := 0
+	callback := func() {
+		for _, state := range states {
+			proxies := tracker.GetProxies(ctx, state.nodeID)
+			require.Equal(t, len(proxies), state.proxiesAfterEachTick[stateCounter])
+		}
+
+		// issue and update for node-2/proxy-3 to keep the relationship alive
+		tracker.AddNode(ctx, input[3].nodeID, input[3].proxyID, input[3].clusterName, input[3].addr)
+
+		wg.Done()
+		stateCounter++
+		if stateCounter < totalStates {
+			clock.Advance(1 * time.Minute)
+		}
+	}
+
+	config := &Config{
+		OfflineThreshold:     1 * time.Minute,
+		Clock:                clock,
+		ProxyControlCallback: callback,
+	}
+	tracker = NewTracker(config)
+
+	for _, node := range input {
+		tracker.AddNode(ctx, node.nodeID, node.proxyID, node.clusterName, node.addr)
+	}
+
+	clock.Advance(1 * time.Minute)
+
+	wg.Wait()
+	tracker.Stop()
+}
+
+// BenchmarkTracker gathers metrics about the tracker
+// running 1,000,000 nodes
+//
+// run the benchmarks using:
+// go test -bench=. -test.cpuprofile=cpu.out -test.memprofile=mem.out
+//
+// inspect the results using
+// go tool pprof implementation.test mem.out
+// go tool pprof implementation.test cpu.out
+//
+// generate a graph
+// go tool pprof --pdf implementation.test mem.out > mem.pdf
+// go tool pprof --pdf implementation.test cpu.out > cpu.pdf
+func BenchmarkTracker(b *testing.B) {
+	b.ReportAllocs()
+
+	ctx := context.Background()
+	config := &Config{OfflineThreshold: 10 * time.Minute}
+	tracker := NewTracker(config)
+
+	input := make(map[string]string, 1000000)
+	for i := 0; i < 1000000; i++ {
+		input[uuid.New()] = uuid.New()
+	}
+
+	// start the benchmark timer.
+	b.ResetTimer()
+
+	// insert 1000000 nodes and proxies relationships
+	// update them 5 times
+	for i := 0; i < 6; i++ {
+		for node, proxy := range input {
+			tracker.AddNode(ctx, node, proxy, "cluster", "proxy:3080")
+		}
+		time.Sleep(10 * time.Second)
+	}
```

Comment:
```suggestion
	for i := 0; i < b.N; i++ {
		for node, proxy := range input {
			tracker.AddNode(ctx, node, proxy, "cluster", "proxy:3080")
		}
	}
```
Benchmarking is a bit of a knowledge gap for me. But this has given me a reason to look into it more. Its unclear why a 10 second sleep is necessary. Maybe my suggestion is sufficient?

Reply:
My issue with b.N is that it's wildly variables based on input, hardware performance/availability...
So it could do 10 iterations on a super fast machine and only 1 on a slow one. You can also change the input vars but but that doesn't make the test consistent across different setups. But that also could be an important factor to consider.

But by hardcoding the values you'll always get performance tests for a set amount of iterations, so you'll be able to compare performance gains with setup improvements. At least that's the way I see it, but I would be down to change it if you feel it's a better way of extracting metrics.

As for the sleep, it doesn't have to be there and it's a random number. It's just a way to simulate the pause between nodes connecting and pinging proxies in real life. Metrics didn't change much for node to proxy relationship updates. if you run the test without the external `for {` loop, you'll basically get the same values as you'd get running it with it.

Reply:
Ah ok that makes sense for the for loop. 

I still think its fine to drop the sleep. I feel like this gives us a good look at our memory usage for storage. If we wanted to get an idea of throughput like requests per second we should add a second test that sets up a server/client and fires off a bunch of requests in parallel.

Reply:
yes doing working on the load tests as we speak.